homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,210 @@
+import homeassistant.components.automation.mqtt as automation_mqtt
+from homeassistant.components.device_automation import TRIGGER_BASE_SCHEMA
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_PLATFORM, CONF_TYPE
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.typing import ConfigType
+
+from . import (
+    ATTR_DISCOVERY_HASH,
+    CONF_CONNECTIONS,
+    CONF_DEVICE,
+    CONF_IDENTIFIERS,
+    CONF_PAYLOAD,
+    CONF_QOS,
+    DOMAIN,
+)
+from .const import DEFAULT_QOS
+from .discovery import MQTT_DISCOVERY_UPDATED, clear_discovery_hash
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AUTOMATION_TYPE = "automation_type"
+CONF_SUBTYPE = "subtype"
+CONF_TOPIC = "topic"
+DEFAULT_ENCODING = "utf-8"
+DEVICE = "device"
+
+MQTT_TRIGGER = {
+    # Trigger when MQTT message is received
+    CONF_PLATFORM: DEVICE,
+    CONF_DOMAIN: DOMAIN,
+}
+
+TRIGGER_SCHEMA = TRIGGER_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Optional(CONF_DEVICE_ID): str,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Optional(CONF_QOS, default=DEFAULT_QOS): vol.Coerce(int),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+PLATFORM_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
+        vol.Required(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    },
+    mqtt.validate_device_has_at_least_one_identifier,
+)
+
+DEVICE_TRIGGERS = "mqtt_device_triggers"
+
+
+@attr.s(slots=True, frozen=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    type = attr.ib(type=str)
+    subtype = attr.ib(type=str)
+    topic = attr.ib(type=str)
+    payload = attr.ib(type=str)
+    qos = attr.ib(type=int)
+    discovery_hash = attr.ib(type=str)
+
+
+async def _update_device(hass, config_entry, config):
+    device_registry = await hass.helpers.device_registry.async_get_registry()
+    config_entry_id = config_entry.entry_id
+    device_info = mqtt.device_info_from_config(config[CONF_DEVICE])
+
+    if config_entry_id is not None and device_info is not None:
+        device_info["config_entry_id"] = config_entry_id
+        device_registry.async_get_or_create(**device_info)
+
+
+def _update_trigger(hass, discovery_hash, config):
+    for device_id, triggers in hass.data[DEVICE_TRIGGERS].items():
+        if discovery_hash in triggers:
```

Comment:
Use guard clause. `if discovery_hash not in triggers: continue`

Reply:
Fixed.